src/lib.rs

```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
+                match container.kind {
+                    PtpContainerType::Data => {
+                        data_phase_payload = payload;
+                    },
+                    PtpContainerType::Response => {
+                        if container.code != StandardResponseCode::Ok {
+                            return Err(Error::Response(container.code));
+                        }
+                        return Ok(data_phase_payload);
+                    },
+                    _ => {}
                 }
+            } else {
+                debug!("mismatched txnid {}, expecting {}", container.tid, tid);
             }
+        }
+    }
 
-            let (mtype, mut msg) = try!(PtpTransaction::parse(&self.buf));
+    // helper for command() above, retrieve container info and payload for the current phase
+    fn read_txn_phase(&mut self) -> Result<(PtpContainerInfo, Vec<u8>), Error> {
+        // large enough to fetch objects other than media in a single read
+        let mut buf = Vec::with_capacity(100 * 1024);
+        let timeout = Duration::from_secs(2);
 
-            if mtype == PtpContainerType::Data && msg.is_response(&transaction) {
-                data = Some(msg.data);
-            } else if mtype == PtpContainerType::Response && msg.is_response(&transaction) {
-                if let Some(data) = data {
-                    msg.data = data;
-                }
-                
-                if msg.code != StandardResponseCode::Ok {
-                    return Err(Error::Response(msg.code));
-                }
-                
-                return Ok(msg.data);
+        let bufslice = unsafe {
+            slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.capacity())
+        };
+        let n = try!(self.handle.read_bulk(self.ep_in, bufslice, timeout));
+        unsafe { buf.set_len(n); }
+
+        let cinfo = try!(PtpContainerInfo::parse(&buf[..]));
+        trace!("container {:?}, bulk rx {}", cinfo, n);
+
+        // response didn't fit into our original buf? resize and read the rest
+        // should ideally require a single read, but loop to handle short reads.
+        while cinfo.payload_len > buf.len() - PTP_CONTAINER_INFO_SIZE {
+            // allocate one extra to deal w/zero length packets appropriately
+            let additional = cinfo.payload_len + PTP_CONTAINER_INFO_SIZE - buf.len() + 1;
+            buf.reserve(additional);
+            let bufslice = unsafe {
+                let p = buf.as_mut_ptr().offset(buf.len() as isize);
+                slice::from_raw_parts_mut(p, buf.capacity() - buf.len())
+            };
+            let n = try!(self.handle.read_bulk(self.ep_in, bufslice, timeout));
+            unsafe {
+                let sz = buf.len();
+                buf.set_len(sz + n);
             }
+            trace!("  bulk rx {}, ({}/{})", n, buf.len(), buf.capacity());
         }
+
+        Ok((cinfo, buf.split_off(PTP_CONTAINER_INFO_SIZE)))
```

Comment:
No, drain still shifts up the part of the buffer after the drained range. It's only an optimization over calling `remove` repeatedly, which shifts for every element removed.

Cutting elements off the beginning of a Vec inherently requires a copy, because the first element and the start of allocation are the same pointer.